include/xtensor/xexpression_traits.hpp

```diff
@@ -128,15 +128,34 @@ namespace xt
             using type = xtensor_fixed<T, xshape<X...>, L>;
         };
     }
-    
-    template <class T, class S, layout_type L>
+
+    template <class Tag, class T>
+    struct temporary_type_from_tag;
+
+    template <class T>
+    struct temporary_type_from_tag<xtensor_expression_tag, T>
+    {
+        using I = std::decay_t<T>;
+        using shape_type = typename I::shape_type;
+        using value_type = typename I::value_type;
+        static constexpr layout_type static_layout = XTENSOR_DEFAULT_LAYOUT;
+        using type = typename detail::xtype_for_shape<shape_type>::template type<value_type, static_layout>;
+    };
+
+    template <class Tag, class T, class = void>
     struct temporary_type
     {
-        using type = typename detail::xtype_for_shape<S>::template type<T, L>;
+        using type = typename temporary_type_from_tag<Tag, T>::type;
+    };
+
+    template <class Tag, class T>
+    struct temporary_type<Tag, T, void_t<typename std::decay_t<T>::temporary_type>>
```

Comment:
Why having the Tag as a template argument instead of retrieving it in the specialization that needs it? This would simplify the call (`using my_type = temporary_type_t<E>;`)

Reply:
I'm not sure it will be enough for other packages like in `xtensor_sparse`.

In `xtensor-sparse`, you will need to call the `temporary_type_t` of  `xtensor_expression_tag` if the result is dense. For me, it will not be possible without adding the tag in the template parameters.

Reply:
This is a detail of implementation of `temporary_type_from_tag<xsparse_expression_tag>`. Nothing prevents you form calling `temporary_type_from_tag<xtensor_expression_tag>` inside it.

Reply:
of course!

Tag removed.